### PR TITLE
Implement simplest demand paging policy

### PR DESCRIPTION
**Pull request by shuyangli**
This implements the simplest demand paging page fault handler. Whenever the system runs into a page fault, our handler allocates a new page, allowing the process to continue.

There exists one problem: physical memory address 0x001a0000 in VirtualBox seems special, in that whenever we touch it, our machine halts. We'll address that in a separate pull request.

Code to test this functionality is under module_tests.h@walk_memory.

### CODE REVIEW

**Comment by dthain** [comment]
Suggest that you add your testing codes to the kernel, and make them invokable from main() so that the user can e.g. type "testpaging" or "testgraphics" to demo those pieces...

File: src/pagefault.c [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    if (pagetable_getmap(current->pagetable, vaddr, &paddr)) {
+        // If it's mapped, then it means we can't access the vaddr.
+        // Dump the process
+        console_printf("interrupt: illegal page access at vaddr %x\n",
+                       vaddr);
+        process_dump(current);
+        process_exit(0);
+    } else {
+        // Otherwise, we know we have a legit page fault
+        // printf("interrupt: page fault at %x\n", vaddr);
```
Comment (kkoser) [inline src/pagefault.c]: pelase remove commented out lines

File: src/pagefault.c [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+        // Currently we give it as much memory as we could
+        pagetable_alloc(current->pagetable, vaddr, PAGE_SIZE,
+            // TODO(SL): figure out if these flags are correct
+            PAGE_FLAG_READWRITE | PAGE_FLAG_USER | PAGE_FLAG_ALLOC);
+
+        // If we can't allocate additional memory, kill the process
+        if (0) {
```
Comment (kkoser) [inline src/pagefault.c]: always false?

File: src/pagetable.c [hunk 3]
```diff
@@ -26,24 +26,38 @@ struct pageentry {
...
+    uint32_t i;
     for (i = 0; i < stop; i += PAGE_SIZE) {
         pagetable_map(p, i, i, PAGE_FLAG_KERNEL | PAGE_FLAG_READWRITE);
     }
-    stop = (unsigned)video_buffer + video_xres * video_yres * 3;
-    for (i = (unsigned)video_buffer; i <= stop; i += PAGE_SIZE) {
+
+    // SL: in VirtualBox, vram is separate from ram, and no matter how much
+    // physical memory it has, video_buffer is always 0xe0000000.
+    // TODO (SL): Ensure video buffer is mapped into superviser mode even
```
Comment (kkoser) [inline src/pagetable.c]: Is there a bug in jira for this? If so we should include the bug number here, ex: NUN-25

File: src/pagefault.c (hunk 2 above)
Comment (shuyangli) [inline src/pagefault.c]: This is for the future, where we'll need to check for memory allocation policy; here we don't have those policies in place yet.

File: src/module_tests.h [hunk 4]
```diff
@@ -0,0 +1,7 @@
+/*
+Copyright (C) 2015 The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file LICENSE for details.
+*/
+
+void walk_memory();
```
Comment (alexmwu) [inline src/module_tests.h:13]: Doxygen docs, please

**Comment by kkoser** [comment]
Squash commits please before merging, you have a lot

File: src/module_tests.c [hunk 5]
```diff
@@ -0,0 +1,31 @@
+/*
```
Comment (kkoser) [inline src/module_tests.c:1]: Can we call this like memory_tests or something more specific?

**Comment by shuyangli** [comment]
@kkoser each commit serves a different purpose, so it makes sense to separate them.

File: src/pagefault.c [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    } else {
+        // Otherwise, we know we have a legit page fault
+        // If we can't allocate additional memory, kill the process
+        if (0) {
+            interrupt_dump_process();
+        }
+
+        // Currently we give it as much memory as we could
+        pagetable_alloc(current->pagetable, vaddr, PAGE_SIZE,
+            // TODO(SL): figure out if these flags are correct
```
Comment (kkoser) [inline src/pagefault.c]: This todo seems kinda worrying to me, in a page fault handler that will be hit a lot, I'd like to know we are using the right flags haha

File: src/pagefault.c [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/*
```
Comment (kkoser) [inline src/pagefault.c]: Do we really need a new file only for the page fault handler? Seems like it should live in the pageable module. All other modules that implement interrupts follow that pattern

File: src/pagetable.h [hunk 8]
```diff
@@ -18,17 +18,109 @@ See the file LICENSE for details.
...
+/**
+ * @brief   Get the physical address of a virtual address in a page directory
+ * @details Given a virtual address, the pagetable gets its corresponding
+ *          physical address. The lookup is successful if the vaddr is mapped
+ *          to a paddr, and the page is currently in memory.
+ *
+ * @param   p       A pointer to the page directory to be looked up
+ * @param   vaddr   Virtual address to be looked up
+ * @param   paddr   A pointer to store the looked up physical address
+ * @return  1 if the lookup is successful, and 0 otherwise
```
Comment (kkoser) [inline src/pagetable.h]: Can we add some discussion here of what a 0 indicates as a return value (here I believe it means that the location is not mapped)

File: src/pagetable.h [hunk 9]
```diff
@@ -18,17 +18,109 @@ See the file LICENSE for details.
...
+ *          virtual page from the given pagetable.
+ *
+ * @param   p       A pointer to the pagetable to be modified
+ * @param   vaddr   A virtual address in the page
+ */
 void pagetable_unmap(struct pagetable *p, unsigned vaddr);
+
+/**
+ * @brief   Allocate pagetables and map a given virtual address and length
+ * @details Given a virtual address, the pagetable unmaps the corresponding
```
Comment (kkoser) [inline src/pagetable.h]: unmaps?

File: src/pagetable.h [hunk 10]
```diff
@@ -18,17 +18,109 @@ See the file LICENSE for details.
...
 void pagetable_alloc(struct pagetable *p, unsigned vaddr, unsigned length,
                      int flags);
+
+/**
+ * @brief   Delete (free) a given pagetable
+ * @details Given a page directory, delete all its contents and free the
+ *          physical memory mapped to the virtual addresses contained in the
+ *          page directory.
+ *
+ * @param   p       A pointer to the pagetable to be deleted
```
Comment (kkoser) [inline src/pagetable.h]: Description indented a lot here

File: src/pagetable.h [hunk 11]
```diff
@@ -18,17 +18,109 @@ See the file LICENSE for details.
...
+ */
 void pagetable_delete(struct pagetable *p);
+
+/**
+ * @brief   Load a given pagetable
+ * @details Load a pagetable into %cr3, so when we swap processes, we can make
+ *          sure the system is using the correct pagetable.
+ *
+ * @param   p   A pointer to the pagetable to be loaded into %cr3
```
Comment (kkoser) [inline src/pagetable.h]: Inconsistent indentation between param name and description

Comment (shuyangli) [inline src/pagetable.h:111]: Can you elaborate? p is the param name

Comment (kkoser) [inline src/pagetable.h:111]: My bad, theres no problem

File: src/module_tests.c [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+Copyright (C) 2015 The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file LICENSE for details.
+*/
+
+#include "console.h"
+#include "pagetable.h"
+#include "process.h"
+
+void walk_memory() {
```
Comment (kkoser) [inline src/module_tests.c:11]: can you change this to be a unit test, and add it to the testing framework?

Comment (kkoser) [inline src/module_tests.c:11]: After discussion, this test is not very quantifiable, as it depends on the length of system memory and will allocate all available memory. It is not really a "unit test"

Comment (alexmwu) [inline src/module_tests.c:11]: Still seems like something useful to have available. Add it to the command line module?

File: src/module_tests.c (hunk 5 above)
Comment (shuyangli) [inline src/module_tests.c:1]: The idea of this file is for any test functions we want temporarily for demo purposes. I'm imagining that after the project has stabilized and the system mostly working, we can get rid of this file in the future in favor of actual OS features.

File: src/module_tests.c (hunk 12 above)
Comment (shuyangli) [inline src/module_tests.c:11]: This function tries to allocate all available memory, so when it finishes the machine will halt; doesn't sound like a good idea to make it readily accessible from the command line, unless we want a sure way to halt the machine

**Comment by jhamilt5** [comment]
A lot to take in, but it looks good to me.

**Comment by shuyangli** [comment]
I'm talking to Jesse and he approves, merging.

**Pull request merged**